Add sync-point insertions and block/thread local memory allocations

zheng-xq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36563 Add sync-point insertions and block/thread local memory allocations**

Differential Revision: [D21014238](https://our.internmc.facebook.com/intern/diff/D21014238)

zheng-xq committed:
Add sync-point insertions and block/thread local memory allocations

[ghstack-poisoned]

zheng-xq committed:
Update on "Add sync-point insertions and block/thread local memory allocations"


Differential Revision: [D21014238](https://our.internmc.facebook.com/intern/diff/D21014238)

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/test_cuda.cpp:
```diff
@@ -574,6 +574,226 @@ void testCudaNoThreadIdxWrite_1() {
+  }
+
+  {
+    //	for m in 0..128:
```
Nit: formatting is off.

zheng-xq replied:
Done

ZolotukhinM commented on test/cpp/tensorexpr/test_cuda.cpp:
```diff
@@ -574,6 +574,226 @@ void testCudaNoThreadIdxWrite_1() {
+  //  for k in 0..1:  // block-idx
+  //    alloc(c, 64)
+  //    for n in 0..64:  // thread-idx
+  //      c(n) = 0
```
For correctness I'd suggest also include `Free` in the test since `Allocate` is present.

Note for myself: IREval should report error if we have unfreed memory in the end of evalutation.

ZolotukhinM commented on test/cpp/tensorexpr/test_cuda.cpp:
```diff
@@ -574,6 +574,226 @@ void testCudaNoThreadIdxWrite_1() {
+  ExpectAllNear(b_v, b_ref, 1e-5);
+
+  cudaFree(a_dev);
+  cudaFree(b_dev);
```
How does this test check the inserted sync-points? Will it fail if we don't insert them?

nickgg replied:
Not consistently I think, bugs would present themselves as race conditions. Another issue would be that the test wont detect over synchronizing (e.g. throwing in a sync between every block).

Can we search the generated cuda for the correct syncthreads?

zheng-xq replied:
The test fails fairly consistently without the functionality. That being said, added a TODO to add the check in a follow-up diff.

ZolotukhinM reviewed (approved):
Looks good, couple of comments are inline!

zheng-xq committed:
Update on "Add sync-point insertions and block/thread local memory allocations"


Differential Revision: [D21014238](https://our.internmc.facebook.com/intern/diff/D21014238)

[ghstack-poisoned]

zheng-xq committed:
Update on "Add sync-point insertions and block/thread local memory allocations"


Differential Revision: [D21014238](https://our.internmc.facebook.com/intern/diff/D21014238)

[ghstack-poisoned]

zheng-xq committed:
Update on "Add sync-point insertions and block/thread local memory allocations"


Differential Revision: [D21014238](https://our.internmc.facebook.com/intern/diff/D21014238)

[ghstack-poisoned]

zheng-xq committed:
Update on "Add sync-point insertions and block/thread local memory allocations"


Differential Revision: [D21014238](https://our.internmc.facebook.com/intern/diff/D21014238)

[ghstack-poisoned]
